src/app/tests/TestICDManager.cpp

```diff
@@ -15,17 +15,204 @@
+        return SUCCESS;
+    }
+
+    static int Finalize(void * context)
+    {
+        auto * ctx = static_cast<TestContext *>(context);
+        ctx->mICDManager.Shutdown();
+        app::EventManagement::DestroyEventManagement();
+        chip::System::Clock::Internal::SetSystemClockForTesting(gRealClock);
+
+        chip::DeviceLayer::PlatformMgr().Shutdown();
+        if (AppContext::Finalize(context) != SUCCESS)
+            return FAILURE;
+
+        return SUCCESS;
+    }
+
+    app::ICDManager mICDManager;
+
+private:
+    TestPersistentStorageDelegate testStorage;
+    MonotonicallyIncreasingCounter<EventNumber> mEventCounter;
+};
+
+} // namespace
+
+namespace chip {
+namespace app {
+class TestICDManager
+{
+public:
+    /*
+     * Advance the test Mock clock time by the amout passed in argument
+     * and then force the SystemLayer Timer event loop. It will check for any expired timer,
+     * and invoke their callbacks if there are any.
+     *
+     * @param time_ms: Value in milliseconds.
+     */
+    static void AdvanceClockAndRunEventLoop(uint32_t time_ms)
+    {
+        LayerSocketsLoop & layer = static_cast<LayerSocketsLoop &>(DeviceLayer::SystemLayer());
+        gMockClock.AdvanceMonotonic(System::Clock::Timeout(time_ms));
+
+        layer.PrepareEvents();
+        layer.WaitForEvents();
+        layer.HandleEvents();
+    }
+
+    static void TestICDModeIntervals(nlTestSuite * aSuite, void * aContext)
+    {
+        TestContext * ctx                   = static_cast<TestContext *>(aContext);
+        Clock::ClockBase * const savedClock = &SystemClock();
+
+        // After the init we should be in active mode
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::ActiveMode);
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeInterval() + 1);
+
+        // Active mode interval expired, ICDManager transitioned to the IdleMode.
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::IdleMode);
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetIdleModeInterval() + 1);
+        // Idle mode interval expired, ICDManager transitioned to the ActiveMode.
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::ActiveMode);
+
+        // Events updating the Operation to Active mode can extend the current active mode time by 1 Active mode threshold.
+        // Kick an active Threshold just before the end of the Active interval and validate that the active mode is extended.
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeInterval() - 1);
+        ctx->mICDManager.UpdateOperationState(ICDManager::OperationalState::ActiveMode);
+
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeThreshold() / 2);
```

Comment:
Missing an NL_ASSERT() check here to confirm the active mode was extended (check we are still in active mode).